[libc][bazel] Add bazel targets for libc/include/... tests.

jtstogel opened the pull request:
This PR also sets `alwayslink=True` for `//libc/test:LibcUnitTest`. This ensures that the `main` function provided always gets linked into a test target. While not strictly necessary, it makes it so tests like https://github.com/llvm/llvm-project/blob/45d8759cbed0f216786729718608a8be72a505c6/libc/test/include/signbit_test.c will give a duplicate symbol error if they incorrectly depend on `//libc/test:LibcUnitTest`.

This PR is missing tests for generated header includes since the current Bazel setup lacks generated headers or a mechanism to run hermetic tests. CMake version of the header include tests: https://github.com/llvm/llvm-project/blob/a2ce5647200ad40ae356affd44db7d054de444d2/libc/test/include/CMakeLists.txt#L515

See issue https://github.com/llvm/llvm-project/issues/134780

vonosmas commented on utils/bazel/llvm-project-overlay/libc/test/UnitTest/BUILD.bazel:
```diff
@@ -71,6 +71,7 @@ libc_test_library(
         "//libc:llvm_libc_macros_stdfix_macros",
         "//llvm:Support",
     ],
+    alwayslink = True,
```
What would happen if you try to link in this library to a test which already has `main()` before this change?

vonosmas replied:
So, both with and without this change the link would fail with `duplicate symbol: main`  if the unit test which has its own main() would depend on `//libc/test/UnitTest:LibcUnitTest`. Could you clarify which behavior are you trying to change?

E.g. if a unit test accidentally has its `int main()` wrapped in a namespace, then both with and without this change the test will instead pick up the main() from the framework, won't it?

jtstogel replied:
> So, both with and without this change the link would fail with duplicate symbol: main if the unit test which has its own `main()` would depend on `//libc/test/UnitTest:LibcUnitTest`.

I could be misconfigured on my machine, but AFAICT this isn't true?

Here's some code examples: https://github.com/llvm/llvm-project/compare/main...jtstogel:llvm-project:linker-opt-example. My understanding (which may be incorrect) is that the linker won't include the object file unless a symbol is used, so if the test has it's own `main()`, the unit test framework main will just be ignored.

> E.g. if a unit test accidentally has its int main() wrapped in a namespace, then both with and without this change the test will instead pick up the main() from the framework, won't it?

Sorry that example wasn't well formed. I think a better argument for why files like these should set `alwayslink=True` is that the program's semantics falls down to link order. In the following example, what does the test do? I would expect it to immediately fail with `duplicate symbol: main`, but on my machine the test either fails or passes based on the ordering of `[":main1", ":main2"]`.

```
cc_test(
  name = "example_test",
  srcs = ["example_test.cc"],
  deps = [":main1", ":main2"]
)

cc_library(name = "main1", srcs = ["main1.cc"])

cc_library(name = "main2", srcs = ["main2.cc"])
```

main1.cc
```
int main(void) { return 1; }
```

main2.cc
```
int main(void) { return 0; }
```

example_test.cc
```
// empty
```

vonosmas replied:
Thanks for the detailed explanation, I agree that this change is reasonable. Would you mind adding a comment above `alwayslink = True` to specify that we deliberately force linking in library to ensure the test case doesn't provide their own `main()`?

vonosmas commented on utils/bazel/llvm-project-overlay/libc/test/include/BUILD.bazel:
```diff
@@ -4,19 +4,315 @@
 
-load("//libc/test:libc_test_rules.bzl", "libc_test")
-
-package(default_visibility = ["//visibility:public"])
```
Could you please restore the default_visibility? It's not too useful for Bazel land, but we're keeping it for consistency.

jtstogel replied:
Yup, my bad, done.

vonosmas commented on utils/bazel/llvm-project-overlay/libc/test/include/sys/BUILD.bazel:
```diff
@@ -0,0 +1,21 @@
+
+licenses(["notice"])
+
+libc_test(
```
I think you can add this target under `libc/test/include/BUILD.bazel` with the rest of the include tests - to be more consistent with CMake setup.

vonosmas commented on utils/bazel/llvm-project-overlay/libc/test/libc_test_rules.bzl:
```diff
@@ -12,36 +12,56 @@ They come in two flavors:
 When performing tests we make sure to always use the internal version.
 """
 
+load("@rules_cc//cc:cc_library.bzl", "cc_library")
```
Is there a reason for migrating away from native.cc_library / native.cc_test in this PR?

jtstogel replied:
Nope, my linter was just complaining, but yeah better not to do so here. Thanks, done.

vonosmas commented on utils/bazel/llvm-project-overlay/libc/test/libc_test_rules.bzl:
```diff
@@ -12,36 +12,56 @@ They come in two flavors:
+        copts = [],
+        deps = [],
+        local_defines = [],
+        use_test_framework = True,
```
I think that instead of adding one more indirection, you may just add `use_test_framework = False` as an optional kwarg to `libc_test`, and specify it everywhere you need, instead of wrapping it in new `libc_c_test` macro. I may very well imagine that for some of the C++ test cases we also wouldn't want to add unit test framework. Conversely, it's not that big of a deal to add `use_test_framework` to all C test targets you're adding.

jtstogel replied:
Sounds good, done.

vonosmas reviewed (approved):
The change looks good after addressing the comment below. Let me know if you'd like to merge this for you.

Thank you for working on this!

jtstogel committed:
Add bazel targets for libc/include/... tests.

These tests are compiled with `-DLIBC_FULL_BUILD`, since otherwise
inclusion of system headers can redefined macros defined by LLVM
libc.

Also, set `alwayslink=True` for `//libc/test:LibcUnitTest`. This ensures that the `main` function provided always gets linked into a test target. While not strictly necessary, it makes it so tests like https://github.com/llvm/llvm-project/blob/45d8759cbed0f216786729718608a8be72a505c6/libc/test/include/signbit_test.c will give a duplicate symbol error if they incorrectly depend on `LibcUnitTest`.

vonosmas reviewed (approved):
LGTM with a couple optional suggestions below.

jtstogel commented on utils/bazel/llvm-project-overlay/libc/test/libc_test_rules.bzl:
```diff
@@ -15,43 +15,83 @@ When performing tests we make sure to always use the internal version.
 
-def libc_test(name, copts = [], deps = [], local_defines = [], **kwargs):
+_FULL_BUILD_LOCAL_DEFINES = [
+    "LIBC_FULL_BUILD",
```
Sounds good, done.
